library/std/src/sys/unix/futex.rs

```diff
@@ -56,31 +104,148 @@ pub fn futex_wait(futex: &AtomicU32, expected: u32, timeout: Option<Duration>) -
+    }
+}
+
+// FreeBSD doesn't tell us how many threads are woken up, so this always returns false.
+#[cfg(target_os = "freebsd")]
+pub fn futex_wake(futex: &AtomicU32) -> bool {
+    use crate::ptr::null_mut;
+    unsafe {
+        libc::_umtx_op(
+            futex as *const AtomicU32 as *mut _,
+            libc::UMTX_OP_WAKE_PRIVATE,
+            1,
+            null_mut(),
+            null_mut(),
+        )
+    };
+    false
+}
+
+#[cfg(target_os = "freebsd")]
+pub fn futex_wake_all(futex: &AtomicU32) {
+    use crate::ptr::null_mut;
+    unsafe {
+        libc::_umtx_op(
+            futex as *const AtomicU32 as *mut _,
+            libc::UMTX_OP_WAKE_PRIVATE,
+            i32::MAX as libc::c_ulong,
+            null_mut(),
+            null_mut(),
+        )
+    };
+}
+
+#[cfg(target_os = "openbsd")]
+pub fn futex_wait(futex: &AtomicU32, expected: u32, timeout: Option<Duration>) -> bool {
+    use crate::convert::TryInto;
+    use crate::ptr::{null, null_mut};
+    let timespec = timeout.and_then(|d| {
+        Some(libc::timespec {
+            // Sleep forever if the timeout is longer than fits in a timespec.
+            tv_sec: d.as_secs().try_into().ok()?,
+            // This conversion never truncates, as subsec_nanos is always <1e9.
+            tv_nsec: d.subsec_nanos() as _,
+        })
+    });
+
+    let r = unsafe {
+        libc::futex(
+            futex as *const AtomicU32 as *mut u32,
+            libc::FUTEX_WAIT,
+            expected as i32,
+            timespec.as_ref().map_or(null(), |t| t as *const libc::timespec),
+            null_mut(),
+        )
+    };
+
+    r == 0 || super::os::errno() != libc::ETIMEDOUT
+}
+
+#[cfg(target_os = "openbsd")]
+pub fn futex_wake(futex: &AtomicU32) -> bool {
+    use crate::ptr::{null, null_mut};
+    unsafe {
+        libc::futex(futex as *const AtomicU32 as *mut u32, libc::FUTEX_WAKE, 1, null(), null_mut())
+            > 0
+    }
+}
+
+#[cfg(target_os = "openbsd")]
 pub fn futex_wake_all(futex: &AtomicU32) {
+    use crate::ptr::{null, null_mut};
     unsafe {
-        libc::syscall(
-            libc::SYS_futex,
-            futex as *const AtomicU32,
-            libc::FUTEX_WAKE | libc::FUTEX_PRIVATE_FLAG,
+        libc::futex(
+            futex as *const AtomicU32 as *mut u32,
+            libc::FUTEX_WAKE,
             i32::MAX,
+            null(),
+            null_mut(),
         );
     }
 }
 
+#[cfg(target_os = "dragonfly")]
+pub fn futex_wait(futex: &AtomicU32, expected: u32, timeout: Option<Duration>) -> bool {
+    use crate::convert::TryFrom;
+    let r = unsafe {
+        libc::umtx_sleep(
+            futex as *const AtomicU32 as *const i32,
+            expected as i32,
+            // A timeout of 0 means infinite, so we round smaller timeouts up to 1 millisecond.
+            // Timeouts larger than i32::MAX milliseconds saturate.
+            timeout.map_or(0, |d| {
+                i32::try_from(d.as_millis()).map_or(i32::MAX, |millis| millis.max(1))
```

Comment:
Wouldn't it make more sense to saturate with 0 for an infinite timeout?

Reply:
I don't think it matters much, but I suppose technically sleeping shorter than requested is correct behaviour, while sleeping longer is not, so I 'round down' / saturate to i32::MAX. (We do have a place or two in `std` where we saturate a `timespec` to an infinite timeout, but reaching i32::MAX milliseconds (≈25 days) is still somewhat realistic, while reaching i32::MAX (≈70 years) or i64::MAX seconds (≈∞ years) is not.)

Reply:
Now that I think about it, saturating to i32::MAX isn't really correct either, since in that case we shouldn't return `false` (`ETIMEDOUT`) when the timeout happens.

So we should either modify the return condition to include this situation as a spurious wakeup rather than a timeout, or just saturate to infinite for huge timeouts. What do you think?

Reply:
Just saturate to infinity: for all sleep operations the general rule is that we sleep *at least* as long as requested, but can sometimes sleep more due to system timer resolution limitations.